File: c++/0739-daily-temperatures/solution.cpp
#include <stack>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int daysCount = temperatures.size();

        vector<int> result(daysCount);

        stack<int> days;
        for (int currentDay = 0; currentDay < daysCount; ++currentDay) {
            while (!days.empty() && temperatures[currentDay] > temperatures[days.top()]) {
                int previousDay = days.top();
                days.pop();

                result[previousDay] = currentDay - previousDay;
            }

            days.push(currentDay);
        }

        return result;
    }
};

File: java/0739-daily-temperatures/Solution.java
import java.util.Stack;

public class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] result = new int[temperatures.length];

        Stack<Integer> days = new Stack<>();
        for (int currentDay = 0; currentDay < temperatures.length; ++currentDay) {
            while (!days.isEmpty() && temperatures[currentDay] > temperatures[days.peek()]) {
                int previousDay = days.pop();
                result[previousDay] = currentDay - previousDay;
            }

            days.add(currentDay);
        }

        return result;
    }
}
